From f7fd71258f15e82ca22ba5d1abee84ffb21f9b20 Mon Sep 17 00:00:00 2001
From: "Friedrich W. H. Kossebau" <kossebau@kde.org>
Date: Sat, 17 Feb 2018 16:54:47 +0100
Subject: [PATCH] [lookandfeel kcm] Do not declare plugin in lookandfeeltool
 code version, v2

Summary:
Also improves race condition on creating the JSON file and running
automoc over the cpp file which refers to it, given there is no
dependency chain defined at all for the lookandfeeltool target and
that kcm_lookandfeel.json, which was prone to make highly parallel
builds fail

Moving the K_PLUGIN_FACTORY_WITH_JSON into a separate file only added
to the sources of kcm_lookandfeel, instead of hiding with an ifndef for
the build of the lookandfeeltool target as tried before, should avoid
any potential confusion of automoc as reported from some builds
(on KDE neon?).

Reviewers: #freebsd, tcberner, bshah, mart, davidedmundson

Subscribers: plasma-devel

Tags: #plasma

Differential Revision: https://phabricator.kde.org/D10607
---
 kcms/lookandfeel/CMakeLists.txt |  2 ++
 kcms/lookandfeel/kcm.cpp        |  7 -------
 kcms/lookandfeel/kcmmain.cpp    | 26 ++++++++++++++++++++++++++
 3 files changed, 28 insertions(+), 7 deletions(-)
 create mode 100644 kcms/lookandfeel/kcmmain.cpp

diff --git a/kcms/lookandfeel/CMakeLists.txt b/kcms/lookandfeel/CMakeLists.txt
index d854eb9d..929ab80c 100644
--- kcms/lookandfeel/CMakeLists.txt
+++ kcms/lookandfeel/CMakeLists.txt
@@ -5,6 +5,7 @@ configure_file (config-kcm.h.cmake ${CMAKE_CURRENT_BINARY_DIR}/config-kcm.h )
 
 
 set(kcm_lookandfeel_SRCS
+  kcmmain.cpp
   kcm.cpp
   ../krdb/krdb.cpp
   ../cursortheme/xcursor/cursortheme.cpp
@@ -62,6 +63,7 @@ kpackage_install_package(package kcm_lookandfeel kcms)
 
 set(lookandfeeltool_SRCS
     lnftool.cpp
+    # TODO: load kcm plugin instead of using code copy
     kcm.cpp
     ../krdb/krdb.cpp
     ../cursortheme/xcursor/cursortheme.cpp
diff --git a/kcms/lookandfeel/kcm.cpp b/kcms/lookandfeel/kcm.cpp
index 58487ca6..ca74367e 100644
--- kcms/lookandfeel/kcm.cpp
+++ kcms/lookandfeel/kcm.cpp
@@ -24,8 +24,6 @@
 #include "config-workspace.h"
 #include <klauncher_iface.h>
 
-#include <KPluginFactory>
-#include <KPluginLoader>
 #include <KAboutData>
 #include <KSharedConfig>
 #include <QDebug>
@@ -57,8 +55,6 @@
 #  include <X11/extensions/Xfixes.h>
 #endif
 
-K_PLUGIN_FACTORY_WITH_JSON(KCMLookandFeelFactory, "kcm_lookandfeel.json", registerPlugin<KCMLookandFeel>();)
-
 KCMLookandFeel::KCMLookandFeel(QObject* parent, const QVariantList& args)
     : KQuickAddons::ConfigModule(parent, args)
     , m_config(QStringLiteral("kdeglobals"))
@@ -813,6 +809,3 @@ bool KCMLookandFeel::resetDefaultLayout() const
 {
     return m_resetDefaultLayout;
 }
-
-
-#include "kcm.moc"
diff --git a/kcms/lookandfeel/kcmmain.cpp b/kcms/lookandfeel/kcmmain.cpp
new file mode 100644
index 00000000..f0bdab25
--- /dev/null
+++ kcms/lookandfeel/kcmmain.cpp
@@ -0,0 +1,26 @@
+/* This file is part of the KDE Project
+   Copyright (c) 2014 Marco Martin <mart@kde.org>
+
+   This library is free software; you can redistribute it and/or
+   modify it under the terms of the GNU Library General Public
+   License version 2 as published by the Free Software Foundation.
+
+   This library is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+   Library General Public License for more details.
+
+   You should have received a copy of the GNU Library General Public License
+   along with this library; see the file COPYING.LIB.  If not, write to
+   the Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
+   Boston, MA 02110-1301, USA.
+*/
+
+#include "kcm.h"
+
+#include <KPluginFactory>
+
+K_PLUGIN_FACTORY_WITH_JSON(KCMLookandFeelFactory, "kcm_lookandfeel.json",
+                           registerPlugin<KCMLookandFeel>();)
+
+#include "kcmmain.moc"

